File: Level-2/insertion_sort.py
#!/usr/bin/python

# Date: 2018-09-08
#
# Description:
# Implement insertion sort.
#
# Approach:
# Scans from right in sorted sub-array and inserts current element at correct
# place by shifting all the elements to right.
#
# Complexity:
# O(n^2)


def insertion_sort_increasing(A):
    for i in range(1, len(A)):
        key = A[i]
        j = i - 1
        while j >= 0 and A[j] > key:
            A[j + 1] = A[j]
            j -= 1
        A[j + 1] = key  # Insert current element at correct place.

def insertion_sort_descending(A):
    for i in range(1, len(A)):  # Fisrt element is sorted in itself
        element = A[i]
        j = i
        while j and A[j - 1] < element:
            A[j] = A[j - 1]
            j -= 1
        A[j] = element

def sort(A):
    B = A[:]
    insertion_sort_increasing(A)
    insertion_sort_descending(B)
    return (A, B)

assert sort([3, 4, 5, 2, 1]) == ([1, 2, 3, 4, 5], [5, 4, 3, 2, 1])
assert sort([3, 4, 5, 2, 1, 6]) == ([1, 2, 3, 4, 5, 6], [6, 5, 4, 3, 2, 1])
assert sort([]) == ([], [])
assert sort([1]) == ([1], [1])
assert sort([2, 1]) == ([1, 2], [2, 1])

File: Level-2/insertion-sort.go
/*
Date: 2023-12-29

Description:
Implement insertion sort.

Approach:
Scans from right in sorted sub-array and inserts current element at correct
place by shifting all the elements to right.

Complexity:
O(n^2)
*/

package main

import (
	"fmt"
	"reflect"
)

// Sorts slice in increasing order using insertion sort.
func insertionSort(nums []int) {
	var j int
	for i := 1; i < len(nums); i++ {
		key := nums[i]
		// Loop until elements are sorted.
		// For decreasing order, reverse the condition.
		for j = i; j > 0 && nums[j-1] > key; j-- {
			nums[j] = nums[j-1]
		}
		nums[j] = key
	}
}

func main() {
	testCases := []struct {
		input, expected []int
	}{
		{input: []int{5, 4, 3, 2, 1, 10, 28, 7, 6}, expected: []int{1, 2, 3, 4, 5, 6, 7, 10, 28}},
		{input: []int{3, 4, 5, 2, 1}, expected: []int{1, 2, 3, 4, 5}},
		{input: []int{3, 4, 5, 2, 1, 6}, expected: []int{1, 2, 3, 4, 5, 6}},
		{input: []int{}, expected: []int{}},
		{input: []int{1}, expected: []int{1}},
		{input: []int{2, 1}, expected: []int{1, 2}},
	}
	for idx, tc := range testCases {
		inputCopy := make([]int, len(tc.input))
		copy(inputCopy, tc.input)
		insertionSort(inputCopy)
		if !reflect.DeepEqual(inputCopy, tc.expected) {
			fmt.Printf("\nTC %d failed. Expected %v, got %v", idx, tc.expected, inputCopy)
		}
	}
}
